lib/smooth/application/display/ILI9341.cpp

```diff
@@ -0,0 +1,399 @@
+    // NOTE: FOR READING PARAMS SCK must be 16MHz or less the datasheet recommends 10MHz.
+    bool ILI9341::read_params(uint8_t cmd, std::vector<uint8_t>& data, uint32_t param_count)
+    {
+        // The spi rx buffer need to be length of multiples of 32 bits to avoid heap corruption.
+        // so we will use DmaFixedBuffer
+        core::util::DmaFixedBuffer<uint8_t, 16> rxdata;
+
+        bool res = false;
+
+        if (param_count == 1)
+        {
+            res = read(cmd, rxdata.data(), 1);
+            data.push_back(rxdata[0]);
+        }
+        else if ((param_count > 1) & (param_count < 16))
+        {
+            // increment param_count by 1 so an additional byte is read and we get all the data
+            res = read(cmd, rxdata.data(), param_count + 1);
+
+            // Since we are using 4-line spi (full duplex) the sck clocks both data out
+            // and data in at the same time.  The first byte going out is the register we want
+            // to read and so the first byte received is garbage or dummy byte.  So we
+            // don't want to copy the dummy byte but copy only the real data
+            for (size_t i = 0; i < param_count; i++)
+            {
+                data.push_back( static_cast<uint8_t>((rxdata[i] << 1) | (rxdata[i + 1] >> 7)));
+            }
+        }
+        else
+        {
+            Log::error(TAG, "read_params() param_count must be greater than 0 or less than 16");
+        }
+
+        return res;
+    }
+
+    // Read a register or a sequence of registers from the display
+    bool ILI9341::read(uint8_t cmd, const uint8_t* rxdata, size_t length)
+    {
+        // if length is 0 nothing to do and probably an error so return
+        bool res = length > 0;
+
+        if (res)
+        {
+            std::lock_guard<std::mutex> lock(get_guard());
+
+            // create 2 spi transactions; one for command and one for reading parameters
+            std::array<spi_transaction_t, 2> trans;
+
+            for (uint8_t x = 0; x < 2; x++)
+            {
+                //Zero out the transaction
+                std::memset(&trans.at(x), 0, sizeof(spi_transaction_t));
+            }
+
+            // reset current transaction counter
+            current_transaction = 0;
+
+            // configure command transaction
+            trans.at(0).rx_buffer = NULL;
+            trans.at(0).rxlength = 0;
+            trans.at(0).length = 8;
+            trans.at(0).tx_data[0] = cmd;
+            trans.at(0).flags = SPI_TRANS_USE_TXDATA;
+
+            // setup pre and post control pin states
+            dc_pretrans_pin_states.at(0) = PIN_LOW;
+            cs_pretrans_pin_states.at(0) = PIN_LOW;
+            cs_posttrans_pin_states.at(0) = PIN_LOW;  // keep chip select low
+
+            // configure read parameters transaction
+            trans.at(1).rx_buffer = const_cast<uint8_t*>(rxdata);
```

Comment:
Avoid cont_cast at all cost. The function signature says it is const uint8_t, but here you say it is not cost, which means you open it up for modification. So will it be modified or not? Who knows? Make the function take a non-const parameter instead or find another solution.